Packs/CiscoUmbrellaReporting/Integrations/CiscoUmbrellaReporting/CiscoUmbrellaReporting.py

```diff
@@ -0,0 +1,1138 @@
+                    "identity_types", "verdict", "file_name", "threats",
+                    "threat_types", "amp_disposition", "page", "page_size"],
+    "intrusion_rule": ["limit", "from", "to", "offset", "signatures", "ip",
+                       "identity_types", "intrusion_action", "ports", "page",
+                       "page_size"]
+}
+
+
+class Client(BaseClient):
+    def __init__(self, base_url: str, organisation_id: str,
+                 secret_key: str, client_key: str,
+                 verify=None,
+                 proxy=None):
+        BaseClient.__init__(
+            self,
+            base_url,
+            verify=verify,
+            proxy=proxy,
+            ok_codes=(200,),
+        )
+        self.token_url = TOKEN_ENDPOINT
+        self.secret_key = secret_key
+        self.client_key = client_key
+        self.organisation_id = int(organisation_id) if \
+            organisation_id.isdigit() else organisation_id
+
+    def access_token(self):
+        """
+        Generate Access token
+        :return: access_token
+        """
+        payload = {
+            "grant_type": 'client_credentials'
+        }
+
+        token_response = requests.post(url=self.token_url, auth=HTTPBasicAuth(
+            username=self.client_key, password=self.secret_key), data=payload)
+        if token_response.status_code == 401:
+            raise DemistoException("Authorization Error: Provided credentials"
+                                   " for Cisco Umbrella Reporting are not valid")
+
+        return token_response.json()['access_token']
+
+    def fetch_data_from_cisco_api(self, end_point: str, params: dict) -> Dict:
+        """
+        : param end_point: Cisco Umbrella Reporting endpoint
+        :return: return the raw api response from Cisco Umbrella Reporting API.
+        """
+        return self.query(end_point, params)
+
+    def query(self, end_point: str, params: dict) -> Dict:
+        """
+
+        :param end_point: Cisco Umbrella Reporting endpoint
+        :param params: Kwargs
+        :return: return the raw api response from Cisco Umbrella Reporting API.
+
+        """
+        result: Dict = {}
+        url_path = f'{self._base_url}/v2/organizations' \
+                   f'/{self.organisation_id}/{end_point}'
+        access_token = self.access_token()
+        payload: Dict = {}
+        response = requests.get(
+            url=url_path,
+            headers={'Authorization': f'Bearer {access_token}'},
+            data=payload,
+            params=params,
+            allow_redirects=False,
+        )
+        if response.status_code in range(300, 310):
+            payload = {}
+            response = requests.get(
+                response.headers['Location'],
+                headers={'Authorization': f'Bearer {access_token}'},
+                data=payload, allow_redirects=True)
+            if response.ok:
+                result = response.json()
+
+        return result
```

Comment:
we have handled now

Reply:
@ShacharKidor  We are checking the success status between 300-310 because Cisco Umbrella Reporting API is redirected to different regions. Yes, it may give 200. We will handle that.
you can refer to [](https://developer.cisco.com/docs/cloud-security/#!api-reference-reports-reporting-overview/http-redirects-and-request-authorization-header)

Reply:
@ShacharKidor 
1. we are checking the status code between 300-310 because Cisco Umbrella Reporting API redirects to different regions. Yes, it can also give 200 so we need to handle this. you can refer to https://developer.cisco.com/docs/cloud-security/#!api-reference-reports-reporting-overview/http-redirects-and-request-authorization-header

Reply:
@ShacharKidor 
1. we are checking the status code between 300-310 because Cisco Umbrella Reporting API redirects to different regions. Yes, it can also give 200 so we need to handle this. For more info, you can refer to the:-
https://developer.cisco.com/docs/cloud-security/#!api-reference-reports-reporting-overview/http-redirects-and-request-authorization-header
2. We will remove the empty payload
3. We will also remove this comment at line 135.
4. We will override the initial response because of the redirection.

Reply:
@RamboV - I'll read about the redirection - still didn't fully understand. 
about 1,2,3 - please correct the code in accordance.

Reply:
@ShacharKidor Change Added

Reply:
Thanks - checking it myself so I'm not marking it as resolved for now.

Reply:
@RamboV - I'm still not sure about the redirection... If we are setting the allow_redirects parameters to True in the first API call, doesn't the cisco API will do the redirect itself? why do we need to add it to our code?

Reply:
@RamboV - I have another correction regarding this function and I apologize that I noticed it only now...
Please use our `self._http_request()` function - instead of calling directly to the requests.get() function. 
For the handling errors part (all the block of `if response.status_code >= 400:`) - you can create a new function called cisco_umbrella_error_handler for example and pass it to the `self._http_request()`.
Please use examples from the content repo (HelloWorld Integration for instance). 
You can reach out to me on slack\zoom tomorrow if you need any help with that.

Reply:
@ShacharKidor modified it

Reply:
@ShacharKidor
As discussed offline if we set allow_redirects=True in the first API call we get Authorization error 401.